### ISSUE TITLE
regex lastIndex

### ISSUE THREAD

**Issue by espretto** [es-shims/es5-shim#403]
the regex `lastIndex` bug seems to persist although _es5shim_ is included.
```js
var empty = ''
var re = /.*/g
re.exec(empty)
re.lastIndex === 0 // false in IE6
```
I've seen this is being worked around to patch String#replace but RegExp#exec is none the wiser.

**Comment by ljharb** [comment]
Indeed, I see that in IE 6, 7, and 8, `var re = /.*/g; alert(re.lastIndex); re.exec(''); alert(re.lastIndex);` alerts 0/1, whereas in IE 9 and everything else, it alerts 0/0.

I'll look into this; thanks for the report!

**Comment by ljharb** [comment]
@espretto where do you see this being worked around with `String#replace`? We're only patching it if the function form reports capture groups incorrectly.

**Comment by espretto** [comment]
my mistake, it wasn't `String#replace` but `String#split` when using regular expressions. [here](https://github.com/es-shims/es5-shim/blob/master/es5-shim.js#L1819) you can find the exact line: `lastIndex = match.index + match[0].length`

**Comment by ljharb** [comment]
Thanks, I'll see what I can do.

This seems like it's only a bug when `exec` is given an empty string, no?

**Comment by espretto** [comment]
unfortunately, i've seen this bug occur on non-empty strings, too. i think the question is whether or not the regex matches *zero* or more of something. for example this yields different results, too:
```js
var re = /[a-z]*/g
re.lastIndex = 1
re.exec('s')
re.lastIndex === 2 // in IE6, expected empty string match at pos 1 instead
re.lastIndex = 1
re.exec('s1')
re.lastIndex === 2 // in IE6, expected empty string match at pos 1 instead
```
for a patch i'm thinking (not tested):
```js
proto.exec = function (str) {
  var match = nativeExec.call(this, str)
  if (match && match[0].length === 0) this.lastIndex -= 1
  return match
}
```
